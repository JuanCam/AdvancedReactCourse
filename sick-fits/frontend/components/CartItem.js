import formatMoney from '../lib/formatMoney';
import styled from 'styled-components';
import PropTyes from 'prop-types';
import RemoveFromCart from './RemoveFromCart';

const CartItemStyled = styled.li`
list-style: none;
padding: 1rem 0;
border-bottom: 1px solid ${props => props.theme.lightgrey};
display: grid;
align-items: center;
grid-template-columns: auto 1fr auto;
img {
    margin-right: 10px;

}
h3 {
    margin: 0;
}
`;

const cartItemComponent = ({ cartItem }) => {
    if (!cartItem.item) return null;
    return (
    <CartItemStyled>
        <img width="100" src={cartItem.item.image} alt="" />
        <div className="cart-item-details">
            <h3>{cartItem.item.title}</h3>
            <p>
            {formatMoney(cartItem.item.price * cartItem.quantity)}
            &ndash;
            <em>{cartItem.quantity} &times; {formatMoney(cartItem.item.price)}</em>
            </p>
        </div>
        <RemoveFromCart id={cartItem.id}/>
    </CartItemStyled>)
};

cartItemComponent.propTyes = {
    cartItemComponent: PropTyes.shape({
        id: PropTyes.number.isRequired,
        quantity: PropTyes.number.isRequired,
    }).isRequired
}

export default cartItemComponent;