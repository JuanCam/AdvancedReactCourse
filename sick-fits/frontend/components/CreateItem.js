import React from 'react';
import Router from 'next/router';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import Error from './ErrorMessage';
import formatMoney from '../lib/formatMoney';

const CREATE_ITEM_MUTATION = gql`
mutation CREATE_ITEM_MUTATION(
        $title: String!
        $description: String!
        $image: String
        $largeImage: String
        $price: Int!
    ) {
    createItem(
        title: $title
        description: $description
        image: $image
        largeImage: $largeImage
        price: $price
    ) {
        id
    }
}
`;

class CreateItem extends React.Component {
    state = {
        title: '',
        description: '',
        image: '',
        largeImage: '',
        price: 0,
    };

    handleChange = ({target}) => {
        const {name, type, value} = target;
        const val = (type === 'number')
        ? parseFloat(value) : value;
        this.setState({[name]: val});
    }

    uploadFile = async ({target}) => {
        const files = target.files;
        const data = new FormData();
        data.append('file', files[0]);
        data.append('upload_preset', 'sickfits');

        const res = await fetch('https://api.cloudinary.com/v1_1/ddxx7sbgs/image/upload', 
        {
            method: 'POST',
            body: data
        });
        const file = await res.json();
        this.setState({
            image: file.secure_url,
            largeImage:  file.eager[0].secure_url
        });
    }

    render() {
        return (
            <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
             {(createItem, {loading, error, called, data}) => (
                    <Form onSubmit={async (e) => {
                        // Stop the form from submiting
                        e.preventDefault();
                        // Call mutation
                        const res = await createItem();
                        // Go to the single item page
                        Router.push({
                            pathname: '/item',
                            query: {id: res.data.createItem.id}
                        })

                    }}>
                    <Error error={error} />
                    
                    <fieldset disabled={loading} aria-bussy={loading}>

                        <label htmlFor="file">Image File</label>
                        <input type="file"
                            id="file"
                            name="file"
                            placeholder="Upload file"
                            required
                            onChange={this.uploadFile} />
                        {this.state.image && <img src={this.state.image} width="150" alt="Upload preview"></img>}

                        <label htmlFor="title">Title</label>
                        <input type="text" 
                        id="title"
                        name="title"
                        placeholder="Title" 
                        required 
                        value={this.state.title}
                        onChange={this.handleChange}/>

                        <label htmlFor="price">Price</label>
                        <input type="number"
                            id="price"
                            name="price"
                            placeholder="Price"
                            required
                            value={this.state.price}
                            onChange={this.handleChange} />

                        <label htmlFor="description">Description</label>
                        <textarea
                            id="description"
                            name="description"
                            placeholder="Enter a description"
                            required
                            value={this.state.description}
                            onChange={this.handleChange} />
                        <button type="submit">Submit</button>
                    </fieldset>
                    </Form>
                )}
            </Mutation>)
    }
}

export default CreateItem;
export { CREATE_ITEM_MUTATION };