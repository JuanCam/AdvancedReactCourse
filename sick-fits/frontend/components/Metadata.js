import Head from 'next/head';

const metaData = () => (
    <Head>
        <meta name="viewport" content="width=device,initial-scale=1"/>
        <meta charSet="utf-8"/>
        <link rel="shortcut icon" href="static/favicon.png"/>
        <link rel="stylesheet" type="text/css" href="static/nprogress.css"/>
        <title>Sick Fits</title>
    </Head>
);

export default metaData;