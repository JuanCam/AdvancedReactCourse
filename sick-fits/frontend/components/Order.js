import React from 'react';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import Head from 'next/head';
import Error from './ErrorMessage';
import OrderStyles from './styles/OrderStyles';
import { Query } from 'react-apollo';
import formatMoney from '../lib/formatMoney';
import { format } from 'date-fns';

const SINGLE_ORDER_QUERY = gql`
query SINGLE_ORDER_QUERY($id: ID!) {
    order(id: $id) {
        id
        charge
        total
        createdAt
        user {
            id
        }
        items {
            id
            title
            description
            price
            image
            quantity
        }
    }
}
`;

class Order extends React.Component {
    static propTypes = {
        id: PropTypes.string.isRequired,
    }

    render() {
        return <Query query={SINGLE_ORDER_QUERY} variables={{ id: this.props.id }}>
            {({data, error, loading}) => {
                if (error) return error;
                if (loading) return <p>Loading...</p>;

                const { order } = data;
                return <div data-test="order"><OrderStyles>
                    <Head>
                        <title>Sick Fits - Order {order.title}</title>
                    </Head>
                    <p>
                        <span>Order ID:</span>
                        <span>{this.props.id}</span>
                    </p>
                    <p>
                        <span>Charge:</span>
                        <span>{order.charge}</span>
                    </p>
                    <p>
                        <span>Date:</span>
                        <span>{format(order.createdAt, 'MMMM d, YYYY h:mm a')}</span>
                    </p>
                    <p>
                        <span>Order Total:</span>
                        <span>{formatMoney(order.total)}</span>
                    </p>
                    <p>
                        <span>Order count:</span>
                        <span>{order.items.length}</span>
                    </p>
                    <div className="items">
                        {order.items.map(item => (<div className='order-item' key={item.id}>
                                {item.name}
                                <img src={item.image} alt={item.title}/>
                                <div className="item-details">
                                    <h2>{item.title}</h2>
                                    <p>Quantity: {item.quantity}</p>
                                    <p>Each: {item.price}</p>
                                    <p>Subtotal: {item.price * item.quantity}</p>
                                    <p>{item.description}</p>
                                </div>
                            </div>))}
                    </div>
                </OrderStyles></div>
            }}
        </Query>;
    }
}

export default Order;
export { SINGLE_ORDER_QUERY };