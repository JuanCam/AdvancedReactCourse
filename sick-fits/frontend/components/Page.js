import React from 'react';
import styled from 'styled-components';
import Header from './Header';
import Metada from './Metadata';

const StyledPage = styled.div`
    background: white;
    color: ${props => props.theme.black};
`;

const Inner = styled.div`
    max-width: ${props => props.theme.maxWidth};
    margin: 0 auto;
    padding: 2rem;
`;

class Page extends React.Component {
     render() {
         return (
         <StyledPage>
             <Metada></Metada>
             <Header />
             <Inner>{this.props.children}</Inner>
         </StyledPage>);
     }
 }

export default Page;