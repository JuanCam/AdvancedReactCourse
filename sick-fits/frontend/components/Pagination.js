import React from 'react';
import PaginationStyles from './styles/PaginationStyles';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import { perPage } from '../config';
import Head from 'next/head';
import Link from 'next/link';

const PAGINATION_QUERY = gql`
query PAGINATION_QUERY {
    itemsConnection {
        aggregate {
            count
        }
    }
}
`;

const pagination = props => (
    
        <Query query={PAGINATION_QUERY}>
            {({data, loading, error}) => {
                if (loading) return <p>Loading!</p>;

                const {count} = data.itemsConnection.aggregate;
                const {page} = props;
                const pages =  Math.ceil(count / perPage);
            return <PaginationStyles data-test="pagination">
                        <Head>
                            <title>
                                Sick Fits | page {page} of {pages} 
                            </title>
                        </Head>
                        <Link prefetch href={{pathname: 'items', query: {page: page - 1}}}>
                        <a className="prev" aria-disabled={page <= 1}>&larr; Prev</a>
                        </Link>
                        <p>{count} Items total</p>
                        <Link prefetch href={{ pathname: 'items', query: { page: page + 1 } }}>
                            <a className="next" aria-disabled={page > pages}>Next &rarr;</a>
                        </Link>
                        <p>Page {page} of <span className="totalPages">{pages}</span></p>
                    </PaginationStyles>
            }}
        </Query>
    
);

export default pagination;
export { PAGINATION_QUERY };