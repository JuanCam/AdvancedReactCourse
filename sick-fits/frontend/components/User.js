import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

const CURRENT_USER_QUERY = gql`
 query {
        me  {
        id
        email
        name
        permissions
    }
 }
`;

const user = props => (
    <Query query={CURRENT_USER_QUERY} {...props} >
        {payload => props.children(payload)}
    </Query>
)

user.PropTypes = {
    children: PropTypes.func.isRequired
};

export default user;
export { CURRENT_USER_QUERY };