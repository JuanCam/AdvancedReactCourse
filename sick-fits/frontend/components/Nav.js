import Link from 'next/link';
import NavStyles from './styles/NavStyles';
import { Mutation } from 'react-apollo';
import User from './User';
import Signout from './Signout';
import CartCount from './CartCount';
import { TOGGLE_CART_MUTATION } from './Cart';

const nav = props => (
    <User>
        {({data: { me }}) => (
            <NavStyles>
                <Link href="/items">
                    <a>Shop</a>
                </Link>
                {me && (
                    <>
                    <Link href="/sell">
                        <a>Sell</a>
                    </Link>
                    <Link href="/orders">
                        <a>Orders</a>
                    </Link>
                    <Link href="/me">
                        <a>Account</a>
                    </Link>
                    <Signout></Signout>
                        <Mutation mutation={TOGGLE_CART_MUTATION}>
                            {toggleCart => {
                                return <button onClick={toggleCart}>
                                    My Cart 
                                    <CartCount 
                                        count={me.cart.reduce((nextItem, cartItem) => nextItem + cartItem.quantity, 0)} />
                                </button>}}
                        </Mutation>
                    </>)}
                {!me && <Link href="/signup">
                    <a>Sign In</a>
                </Link>
                }
            </NavStyles>
        )}
    </User>
    );

export default nav;