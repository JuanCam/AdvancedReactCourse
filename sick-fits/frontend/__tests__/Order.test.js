import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import Order, { SINGLE_ORDER_QUERY } from '../components/Order';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser, fakeOrder } from '../lib/testUtils';
import { ApolloConsumer } from 'react-apollo';

const order = fakeOrder();
const mocks = [{
    request: {
        query: SINGLE_ORDER_QUERY,
        variables: { id: 'ord123' }
    },
    result: {
        data: {
            order
        }
    }
}];

describe('<Order />', () => {

    it('renders and matches snapshot', async () => {
        const wrapper = mount(<MockedProvider mocks={mocks}>
            <Order id={order.id} />
        </MockedProvider>);

        await wait();
        wrapper.update();
        expect(toJSON(wrapper.find('[data-test="order"]'))).toMatchSnapshot();
    });

    it('should render the items properly', async () => {
        const wrapper = mount(<MockedProvider mocks={mocks}>
            <Order id={order.id} />
        </MockedProvider>);

        await wait();
        wrapper.update();
        expect(wrapper.find('.items .order-item')).toHaveLength(2);
    });

});