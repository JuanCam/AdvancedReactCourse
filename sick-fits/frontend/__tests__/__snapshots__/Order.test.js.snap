// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order /> renders and matches snapshot 1`] = `
<div
  data-test="order"
>
  <OrderStyles>
    <div
      className="OrderStyles-sc-4oqalm-0 camMij"
    >
      <SideEffect(Head)>
        <Head />
      </SideEffect(Head)>
      <p>
        <span>
          Order ID:
        </span>
        <span>
          ord123
        </span>
      </p>
      <p>
        <span>
          Charge:
        </span>
        <span>
          ch_123
        </span>
      </p>
      <p>
        <span>
          Date:
        </span>
        <span>
          March 31, 2018 7:00 PM
        </span>
      </p>
      <p>
        <span>
          Order Total:
        </span>
        <span>
          $400
        </span>
      </p>
      <p>
        <span>
          Order count:
        </span>
        <span>
          2
        </span>
      </p>
      <div
        className="items"
      >
        <div
          className="order-item"
          key="f9bcce46-b516-4189-95ff-fa6ae8c415c3"
        >
          <img
            alt="amet modi reprehenderit omnis adipisci eius laborum"
            src="tenetur.jpg"
          />
          <div
            className="item-details"
          >
            <h2>
              amet modi reprehenderit omnis adipisci eius laborum
            </h2>
            <p>
              Quantity: 
              1
            </p>
            <p>
              Each: 
              4234
            </p>
            <p>
              Subtotal: 
              4234
            </p>
            <p>
              eius aut illo dolorum ab aliquam id
            </p>
          </div>
        </div>
        <div
          className="order-item"
          key="3fd03e80-9d9d-4810-a8de-fc8184c94bf7"
        >
          <img
            alt="assumenda quasi adipisci mollitia iste itaque omnis"
            src="molestias.jpg"
          />
          <div
            className="item-details"
          >
            <h2>
              assumenda quasi adipisci mollitia iste itaque omnis
            </h2>
            <p>
              Quantity: 
              1
            </p>
            <p>
              Each: 
              4234
            </p>
            <p>
              Subtotal: 
              4234
            </p>
            <p>
              debitis rem similique temporibus aliquam sed sapiente
            </p>
          </div>
        </div>
      </div>
    </div>
  </OrderStyles>
</div>
`;
